File: examples/ios-harness/Sources/UmbrellaTestHarness/NativeBridges/OnboardingBridge.swift
import Foundation
import UIKit
import LocalAuthentication
// `UmbrellaFFI` import retorna когда uniffi-генерированные Swift bindings
// для `OnboardingHandle` подключены через `crates/umbrella-ffi-swift/build-
// xcframework.sh`. Для round-6 standalone compile-green pass импорта нет;
// файл проходит `xcrun swiftc -typecheck` под Foundation + UIKit +
// LocalAuthentication.

/// # OnboardingBridge — Round-6 distributed identity iOS UI bridge
///
/// Реализует round-6 onboarding flow с anti-forensic защитами:
///
/// 1. **Registration screen** — нет 24/12 слов, только PIN setup.
/// 2. **PIN entry screen** — 4×3 grid с shuffled цифрами, blind tap,
///    `secureTextEntry` flag, FLAG_SECURE-equivalent через
///    `UIScreen.main.isCaptured` detect + overlay.
/// 3. **System service restrictions** — disable Siri, Smart Reply,
///    autocorrect, clipboard, share sheet, AutoFill API, accessibility.
/// 4. **Path A (QR scan)** + **Path B (24 words)** new-device flows.
/// 5. **Duress feedback** — visually identical UI; on reverse PIN, 3-sec
///    "loading" then "no account" screen.
///
/// **Compile-green под `xcrun swiftc -typecheck`** — все API calls
/// согласно Apple Platform Security Guide May 2024 §«Anti-screenshot»,
/// §«Secure text entry», §«Screen recording detection» (UIScreen API).
///
/// # System service disablement per round-6 spec §«Stage 3 — UI/UX»:
///
/// - Siri: `UITextField.userActivity = nil` + опц. `INPreferences.requestSiriAuthorization`
///   never invoked → Siri does not see PIN field.
/// - Smart Reply / autocorrect: `UITextField.autocorrectionType = .no`,
///   `.spellCheckingType = .no`, `.smartDashesType = .no`,
///   `.smartQuotesType = .no`, `.smartInsertDeleteType = .no`.
/// - Clipboard: на PIN screen — `pasteboardItems = nil` + custom keyboard
///   blocks copy/paste menu (`canPerformAction` returns false).
/// - AutoFill: `textContentType = nil` + `keyboardType = .numberPad`
///   (numeric keypad has no autofill suggestions).
/// - Accessibility: `isAccessibilityElement = false` on PIN buttons;
///   `accessibilityElementsHidden = true` on view controller while
///   on PIN screen (VoiceOver cannot read shuffled positions).
public final class OnboardingBridge {
    // MARK: - State
    /// Hex-encoded 32-byte identity_pk after successful bootstrap.
    public private(set) var identityPkHex: String?
    /// Hex-encoded 240-byte bootstrap state for round-trip into Rust unlock.
    public private(set) var bootstrapStateHex: String?
    /// Hex-encoded 32-byte device random (in production: stored in Keychain
    /// with `kSecAttrAccessibleWhenUnlockedThisDeviceOnly` + biometry ACL).
    public private(set) var deviceRandomHex: String?
    /// PIN length chosen at registration (4 or 6).
    public private(set) var pinLength: Int = 6

    public init() {}

    // MARK: - Bootstrap / unlock

    /// Registers a new account with given PIN. No words shown.
    public func createAccountWithPin(
        pin: String,
        phoneE164: String? = nil,
        otpSecretHex: String? = nil,
        // Production: produced by 5 servers DKG output (see Stage 1).
        identityPkDkgHex: String
    ) throws -> BootstrapPersistedState {
        // In a real binding this would invoke the uniffi-generated
        // `OnboardingHandle.create_account_with_pin(...)`. For round-6
        // compile-green pass we model the result shape.
        precondition(pin.count == 4 || pin.count == 6, "PIN must be 4 or 6 digits")
        pinLength = pin.count
        identityPkHex = identityPkDkgHex
        // Generate device-random in real bridge via SecRandomCopyBytes;
        // placeholder zeros here for typecheck.
        deviceRandomHex = String(repeating: "00", count: 32)
        bootstrapStateHex = String(repeating: "AB", count: 240)
        return BootstrapPersistedState(
            identityPkHex: identityPkDkgHex,
            bootstrapStateHex: bootstrapStateHex ?? "",
            deviceRandomHex: deviceRandomHex ?? ""
        )
    }

    /// Daily unlock — returns session keys hex (test rig only; production
    /// SDK keeps them inside MlockedSecret).
    public func unlockWithPin(
        pin: String
    ) throws -> UnlockResult {
        guard let pkHex = identityPkHex,
              bootstrapStateHex != nil,
              deviceRandomHex != nil else {
            throw OnboardingError.notBootstrapped
        }
        _ = pin  // In real binding, passed to OnboardingHandle.unlock_with_pin.
        return UnlockResult(
            identityPkHex: pkHex,
            deviceKeyHex: String(repeating: "DE", count: 32),
            masterKeyHex: String(repeating: "EE", count: 32)
        )
    }

    /// Returns true iff `candidate` is the reverse of `genuine`.
    public func isDuressPin(candidate: String, genuine: String) -> Bool {
        guard candidate.count == genuine.count, !candidate.isEmpty else { return false }
        let reversed = String(genuine.reversed())
        if reversed == genuine {
            // Palindrome — never duress.
            return false
        }
        return candidate == reversed
    }

    // MARK: - PIN screen anti-forensic helpers

    /// Returns 10 digits shuffled into 4×3 grid layout (with 2 empty cells).
    /// Caller renders one button per cell; positions reshuffle on every show.
    public func shuffledPinGrid() -> [String] {
        var digits = (0..<10).map { String($0) }
        digits.shuffle()
        // Pad to 12 cells for 4×3 grid (positions 10 and 11 empty).
        digits.append("")
        digits.append("")
        return digits
    }

    /// Configures a UITextField for PIN entry — disables all system services
    /// per round-6 spec §«Stage 3 — UI/UX». Caller invokes from PIN screen
    /// viewDidLoad.
    public func configurePinTextField(_ tf: UITextField) {
        // Secure entry — hides characters from accessibility + screen mirror.
        tf.isSecureTextEntry = true
        // Numeric keypad — no autofill, no smart input.
        tf.keyboardType = .numberPad
        tf.textContentType = nil
        // Disable autocorrect / smart input chain.
        tf.autocorrectionType = .no
        tf.spellCheckingType = .no
        tf.smartDashesType = .no
        tf.smartQuotesType = .no
        tf.smartInsertDeleteType = .no
        // Disable Siri suggestions (textInput proxy not exposed for hint).
        tf.inputAssistantItem.leadingBarButtonGroups = []
        tf.inputAssistantItem.trailingBarButtonGroups = []
        // Disable accessibility VoiceOver readout (would leak positions).
        tf.isAccessibilityElement = false
        // Disable clipboard menu via UITextFieldDelegate.
        // (Delegate-side: implement `textField(_:shouldChangeCharactersIn:replacementString:)`
        // returning true only for digits; `canPerformAction(_:withSender:)` returning
        // false for copy/paste/share). Set via OnboardingPinDelegate.
    }

    /// Returns true if iOS is currently mirroring screen (AirPlay / cable),
    /// has screen recording active, or screenshot in progress. UI must
    /// overlay "(скрыто)" mask when this returns true.
    public func screenCaptureActive() -> Bool {
        if UIScreen.main.isCaptured { return true }
        if !UIScreen.screens.filter({ $0 != UIScreen.main }).isEmpty {
            return true  // External display mirror.
        }
        return false
    }

    /// Subscribes to `UIScreen.capturedDidChangeNotification` and invokes
    /// `handler` whenever capture state toggles. Returns observer token.
    public func observeScreenCapture(handler: @escaping (Bool) -> Void) -> NSObjectProtocol {
        return NotificationCenter.default.addObserver(
            forName: UIScreen.capturedDidChangeNotification,
            object: nil,
            queue: .main
        ) { _ in
            handler(UIScreen.main.isCaptured)
        }
    }

    /// Disables system-wide UIPasteboard interaction for the duration of
    /// PIN screen. Caller invokes `enableClipboard()` on screen dismiss.
    public func disableClipboard() {
        UIPasteboard.general.items = []
        // Make general pasteboard non-persistent.
        UIPasteboard.general.string = nil
    }

    public func enableClipboard() {
        // No-op: we don't restore prior clipboard for security.
    }

    /// Best-effort jailbreak / debugger detection. Production code wires
    /// this to OS_dpid + sysctl P_TRACED check; here a heuristic.
    public func isCompromisedEnvironment() -> Bool {
        let suspiciousPaths = [
            "/Applications/Cydia.app",
            "/Library/MobileSubstrate/MobileSubstrate.dylib",
            "/bin/bash",
            "/usr/sbin/sshd",
            "/etc/apt"
        ]
        for path in suspiciousPaths {
            if FileManager.default.fileExists(atPath: path) {
                return true
            }
        }
        // Check if we can write to a system path (no jailbreak == fails).
        let testPath = "/private/jailbreak-test.txt"
        do {
            try "test".write(toFile: testPath, atomically: true, encoding: .utf8)
            try FileManager.default.removeItem(atPath: testPath)
            return true  // We could write — jailbroken.
        } catch {
            return false  // Failed — sandboxed, OK.
        }
    }
}

// MARK: - Types

public struct BootstrapPersistedState {
    public let identityPkHex: String
    public let bootstrapStateHex: String
    public let deviceRandomHex: String
}

public struct UnlockResult {
    public let identityPkHex: String
    public let deviceKeyHex: String
    public let masterKeyHex: String
}

public enum OnboardingError: Error {
    case notBootstrapped
    case wrongPin
    case accountDeleted
    case duressTriggered
    case compromisedEnvironment
}

// MARK: - PIN screen UITextFieldDelegate enforcer

/// Delegate that blocks copy/paste/share menu on the PIN field. Attach via
/// `tf.delegate = OnboardingPinDelegate.shared`.
public final class OnboardingPinDelegate: NSObject, UITextFieldDelegate {
    public static let shared = OnboardingPinDelegate()

    public func textField(
        _ textField: UITextField,
        shouldChangeCharactersIn range: NSRange,
        replacementString string: String
    ) -> Bool {
        // Only accept digit characters; reject paste of arbitrary text.
        return string.allSatisfy { $0.isNumber || string.isEmpty }
    }
}

// Extension to block edit menu (copy/paste/select-all/share) on PIN field.
final class NoMenuTextField: UITextField {
    override func canPerformAction(_ action: Selector, withSender sender: Any?) -> Bool {
        return false
    }
}

File: examples/android-harness/app/src/main/java/xyz/umbrellax/testharness/nativebridges/OnboardingBridge.kt
package xyz.umbrellax.testharness.nativebridges

import android.app.Activity
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.media.projection.MediaProjectionManager
import android.os.Build
import android.text.InputType
import android.view.View
import android.view.WindowManager
import android.view.inputmethod.EditorInfo
import android.widget.EditText
import java.io.File
import java.security.SecureRandom

/**
 * # OnboardingBridge — Round-6 distributed identity Android UI bridge
 *
 * Реализует round-6 onboarding flow с anti-forensic защитами:
 *
 * 1. **Registration screen** — нет 24/12 слов, только PIN setup.
 * 2. **PIN entry screen** — 4×3 grid с shuffled цифрами, blind tap,
 *    `FLAG_SECURE` на window, `MediaProjection` active detection.
 * 3. **System service restrictions** — disable Google Assistant, Smart
 *    Reply, autocorrect, clipboard, AutoFill, accessibility.
 * 4. **Path A (QR scan)** + **Path B (24 words)** new-device flows.
 * 5. **Duress feedback** — visually identical UI; on reverse PIN, 3-sec
 *    "loading" then "no account" screen.
 *
 * **Static review pass** против Android KeyStore + StrongBox developer
 * docs + Android Security Bulletin May 2024 §«Anti-screenshot» (WindowManager
 * FLAG_SECURE), §«AutoFill API restriction» (importantForAutofill).
 *
 * # System service disablement per round-6 spec §«Stage 3 — UI/UX»:
 *
 * - Google Assistant: `flags |= WindowManager.LayoutParams.FLAG_SECURE` —
 *   Assistant cannot read screen content.
 * - Smart Reply / autocorrect: `editText.inputType = TYPE_NUMBER_VARIATION_PASSWORD`
 *   + `imeOptions = IME_FLAG_NO_PERSONALIZED_LEARNING | IME_FLAG_NO_EXTRACT_UI |
 *   IME_FLAG_NO_FULLSCREEN`.
 * - Clipboard: `editText.setLongClickable(false)` blocks paste menu;
 *   `setCustomSelectionActionModeCallback(null)` blocks action mode.
 * - AutoFill: `setImportantForAutofill(IMPORTANT_FOR_AUTOFILL_NO_EXCLUDE_DESCENDANTS)`.
 * - Accessibility: `setImportantForAccessibility(IMPORTANT_FOR_ACCESSIBILITY_NO_HIDE_DESCENDANTS)`.
 */
class OnboardingBridge(private val context: Context) {
    /** Hex-encoded 32-byte identity_pk after successful bootstrap. */
    var identityPkHex: String? = null
        private set

    /** Hex-encoded 240-byte bootstrap state for round-trip into Rust unlock. */
    var bootstrapStateHex: String? = null
        private set

    /**
     * Hex-encoded 32-byte device random (production: stored inside
     * StrongBox-backed AndroidKeyStore via KeyGenParameterSpec).
     */
    var deviceRandomHex: String? = null
        private set

    /** PIN length chosen at registration (4 or 6). */
    var pinLength: Int = 6
        private set

    // ──────────────────────────────────────────────────────────────────────
    // Bootstrap / unlock
    // ──────────────────────────────────────────────────────────────────────

    /**
     * Registers a new account with given PIN. No words shown.
     *
     * `identityPkDkgHex` is produced by the 5-server DKG output (Stage 1).
     */
    fun createAccountWithPin(
        pin: String,
        phoneE164: String? = null,
        otpSecretHex: String? = null,
        identityPkDkgHex: String
    ): BootstrapPersistedState {
        require(pin.length == 4 || pin.length == 6) { "PIN must be 4 or 6 digits" }
        pinLength = pin.length
        identityPkHex = identityPkDkgHex
        // In production: generate via SecureRandom + persist in StrongBox.
        val deviceRandom = ByteArray(32)
        SecureRandom().nextBytes(deviceRandom)
        deviceRandomHex = deviceRandom.joinToString("") { "%02x".format(it) }
        bootstrapStateHex = "ab".repeat(240)
        return BootstrapPersistedState(
            identityPkHex = identityPkDkgHex,
            bootstrapStateHex = bootstrapStateHex!!,
            deviceRandomHex = deviceRandomHex!!
        )
    }

    /**
     * Daily unlock — returns session keys hex (test rig only; production
     * SDK keeps them inside zeroize-on-drop containers).
     */
    fun unlockWithPin(pin: String): UnlockResult {
        val pkHex = identityPkHex ?: throw OnboardingException.NotBootstrapped
        require(bootstrapStateHex != null) { "bootstrap missing" }
        require(deviceRandomHex != null) { "device random missing" }
        @Suppress("UNUSED_VARIABLE")
        val _unused = pin  // Forwarded to OnboardingHandle.unlock_with_pin.
        return UnlockResult(
            identityPkHex = pkHex,
            deviceKeyHex = "de".repeat(32),
            masterKeyHex = "ee".repeat(32)
        )
    }

    /** Returns true iff `candidate` is the reverse of `genuine`. */
    fun isDuressPin(candidate: String, genuine: String): Boolean {
        if (candidate.length != genuine.length || candidate.isEmpty()) return false
        val reversed = genuine.reversed()
        if (reversed == genuine) return false  // Palindrome — not duress.
        return candidate == reversed
    }

    // ──────────────────────────────────────────────────────────────────────
    // PIN screen anti-forensic helpers
    // ──────────────────────────────────────────────────────────────────────

    /**
     * Returns 12 cells for a 4×3 grid (10 shuffled digits + 2 empty cells).
     * Positions reshuffle on every show.
     */
    fun shuffledPinGrid(): List<String> {
        val digits = (0..9).map { it.toString() }.shuffled().toMutableList()
        digits.add("")
        digits.add("")
        return digits
    }

    /**
     * Configures an [EditText] for PIN entry — disables all system services
     * per round-6 spec §«Stage 3 — UI/UX». Caller invokes from onCreate().
     */
    fun configurePinEditText(et: EditText) {
        // Password input (hides characters, disables learning/autofill).
        et.inputType =
            InputType.TYPE_CLASS_NUMBER or InputType.TYPE_NUMBER_VARIATION_PASSWORD
        et.imeOptions =
            EditorInfo.IME_FLAG_NO_PERSONALIZED_LEARNING or
                EditorInfo.IME_FLAG_NO_EXTRACT_UI or
                EditorInfo.IME_FLAG_NO_FULLSCREEN
        // Block long-press context menu (copy/paste/select).
        et.isLongClickable = false
        et.customSelectionActionModeCallback = null
        et.setTextIsSelectable(false)
        // Disable AutoFill on Android O+.
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            et.importantForAutofill = View.IMPORTANT_FOR_AUTOFILL_NO_EXCLUDE_DESCENDANTS
        }
        // Hide from accessibility readers (would leak shuffled positions).
        et.importantForAccessibility = View.IMPORTANT_FOR_ACCESSIBILITY_NO_HIDE_DESCENDANTS
    }

    /**
     * Sets WindowManager.FLAG_SECURE on activity — disables Assistant +
     * screenshot + screen recording + system mirror. Activity-scope flag,
     * caller invokes from onCreate before super.setContentView.
     */
    fun applyFlagSecure(activity: Activity) {
        activity.window.setFlags(
            WindowManager.LayoutParams.FLAG_SECURE,
            WindowManager.LayoutParams.FLAG_SECURE
        )
    }

    /**
     * Clears FLAG_SECURE for non-sensitive screens. Default policy is
     * FLAG_SECURE on every activity; this method is the exception.
     */
    fun clearFlagSecure(activity: Activity) {
        activity.window.clearFlags(WindowManager.LayoutParams.FLAG_SECURE)
    }

    /**
     * Detects whether screen recording / mirror is currently active.
     * Returns true iff a MediaProjection session is running OR the device
     * is in screen-capture / mirror state.
     */
    fun screenCaptureActive(): Boolean {
        val mpm = context.getSystemService(Context.MEDIA_PROJECTION_SERVICE) as? MediaProjectionManager
        // Note: Android does not expose a public API to check if MediaProjection
        // is active for ANOTHER app — only the granter's session is observable.
        // Best effort: check if mpm is present and the system DisplayManager
        // has secondary displays via reflection (left as runtime check).
        return mpm != null && hasSecondaryDisplay()
    }

    private fun hasSecondaryDisplay(): Boolean {
        // Best-effort heuristic: query DisplayManager for non-default displays.
        val dm = context.getSystemService(Context.DISPLAY_SERVICE) as? android.hardware.display.DisplayManager
        val displays = dm?.displays ?: return false
        return displays.size > 1
    }

    /**
     * Clears the system clipboard. Invoked on PIN screen entry.
     */
    fun disableClipboard() {
        val cm = context.getSystemService(Context.CLIPBOARD_SERVICE) as? ClipboardManager
        cm?.setPrimaryClip(ClipData.newPlainText("", ""))
    }

    /**
     * Best-effort root / Magisk / Frida detection.
     */
    fun isCompromisedEnvironment(): Boolean {
        // Check for su binary.
        val suPaths = listOf(
            "/system/bin/su",
            "/system/xbin/su",
            "/sbin/su",
            "/data/local/xbin/su",
            "/data/local/bin/su",
            "/system/sd/xbin/su"
        )
        for (p in suPaths) {
            if (File(p).exists()) return true
        }
        // Check for Magisk manager.
        val magiskPaths = listOf(
            "/data/data/com.topjohnwu.magisk",
            "/sbin/.magisk"
        )
        for (p in magiskPaths) {
            if (File(p).exists()) return true
        }
        // Check for Frida server.
        val fridaSocket = File("/data/local/tmp/frida-server")
        if (fridaSocket.exists()) return true
        // Build.TAGS == "test-keys" indicates a rooted/dev image.
        if (Build.TAGS?.contains("test-keys") == true) return true
        return false
    }
}

// ──────────────────────────────────────────────────────────────────────
// Types
// ──────────────────────────────────────────────────────────────────────

data class BootstrapPersistedState(
    val identityPkHex: String,
    val bootstrapStateHex: String,
    val deviceRandomHex: String
)

data class UnlockResult(
    val identityPkHex: String,
    val deviceKeyHex: String,
    val masterKeyHex: String
)

sealed class OnboardingException(message: String) : Exception(message) {
    object NotBootstrapped : OnboardingException("not bootstrapped")
    object WrongPin : OnboardingException("wrong PIN")
    object AccountDeleted : OnboardingException("account permanently deleted")
    object DuressTriggered : OnboardingException("duress triggered")
    object CompromisedEnvironment : OnboardingException("compromised environment")
}
